dev/camdenhells/REQUIREMENTS.md

```diff
@@ -324,41 +350,75 @@ This does not preclude performance impacts; clearly there will be limits, but ha
 3. Provide "successful completion" callback
-  * Provide a variant that includes queueing to main/UI thread ? 
-
-# 4. Testing, Debugging, and Reliability Requirements
-
-1. Automated unit/integration test harness and tests 
-  * Details ?
-2. Internal runtime logging/tracing/assertions?
-  * E.g. for integration testing and for debugging support issues
-3. Compile-time API-usage validation, e.g. annotations ?
-  * Are these enforceable e.g. via lint?
-4. Runtime API-usage validation for the app, e.g. debug-only assertions/exceptions?
-  * 
-5. Debug visualizations (#1436)
-  * Trident, camera eyeball/frustum, lighting rays/lines, object bounds
-  * Are there (non-debug) app use cases for some of these as well? e.g.: 
-    * Interactive camera/lighting selections from a global-view PIP inset 
-    * Theater/stage lighting design apps
-    * Interior design apps - door/window/furniture/light placements
-  * If so perhaps just call them e.g. camera/light/bounds visualizations  
+  * Provide a variant that includes queueing to main/UI thread (see above)
+
+# 4. Testing, Validation, and Debugging Requirements
+
+## 4.1 Compile-time
+
+  * Consistently specify any constraints on all member and parameter values
+    * Specify all references as either nullable or non-null
+    * Specify integer and float ranges whenever less than built-in language limits
+  * ~Validate public API-usage [2.1]~
+     * ~Create an Android lint plugin to detect invocations of non API methods?~
+     * Seems like a reliable library-specific tool would be a lot of work/maintenance, and unlikely that client developers would download/install/use it anyway
+
+## 4.2 Build-time 
+
+  Help! I'm trying to define the line on what should be unit tested and what should not, and don't even understand the hardware/software environment in which unit tests run (i.e. emulators use what version of OpenGL? is there an Android process/activity context? etc)
+  
+  * Automated unit/integration test harness and tests 
+  * For all public API and internal interface methods except those:
+    * With significant historical runtime event dependencies or behavioral states, and/or
+    * Whose success cannot be (at least partially) reasonably determined as value assertions
+  * Includes e.g. contructors, simple get/set accessors, deterministic functions/transforms
+  * Excludes e.g. frame draws, client callbacks
+  
+## 4.3. Run-time
+
+For engine and app development and integration testing, and for debugging support issues
+
+1. Provide unified logging
+  * The usual levels - verbose, debug, info, warning, error, wtf
+  * All messages should identify the class and method
+  * Core engine/rendering component (scenes, scene view, render pass, etc.) messages should also identify the instance
+  * All warning and error messages should include specific causes and data values if appropriate
+  * All error messages should be accompanied by a thrown exception
```

Comment:
Unless rendering can continue. For example, a singular program failing to compile is definitely an error, however rendering can continue with other programs. Only parts of the render using the failed program will not display.

Reply:
Doesn't that imply some fairly sophisticated runtime logic to decide if rendering can continue and how to partially disable the bad parts (e.g. this object has a bad program/VBO/whatever so skip rendering it)?  I.e. should we use error to mean unrecoverable, and warning to mean we think we can keep going (without any extra logic)?

Reply:
Its pretty straight forward and we have them in place now (1.x). If the VBO handles are invalid or the program handles are invalid, we skip that render call and move to the next. I am also ok with the idea of this being a warning and keeping the errors accompanied by exceptions

Reply:
OK, will add the caveat for GL resource warnings and leave errors with exceptions